### PR TITLE
[vulkan] Add vulkan profiler

### PR DESCRIPTION
**Pull request by turbo0628**
Add profiler for vulkan backend.

Known issues:
* Apple M1 cannot run over 32 concurrent query pools, it requires frequent sync so the performance data is inaccurate.

### CODE REVIEW

File: taichi/rhi/device.h [hunk 1]
```diff
@@ -10,6 +10,8 @@
 #include "taichi/rhi/device_capability.h"
 #include "taichi/rhi/arch.h"
 
+#include "taichi/program/kernel_profiler.h"
```
Comment (bobcao3) [inline taichi/rhi/device.h:13]: `device.h` should not include any taichi headers

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

File: taichi/rhi/device.h (hunk 1 above)
Comment (turbo0628) [reply taichi/rhi/device.h]: Fixed it :)

File: taichi/rhi/device.h [hunk 2]
```diff
@@ -429,6 +431,16 @@ class TI_DLL_EXPORT CommandList {
     return RhiResult::not_supported;
   }
 
+  // Profiler support
+  virtual std::unique_ptr<KernelProfilerSamplingHandlerBase>
```
[the remark this answers is not in the document]
Comment (turbo0628) [reply taichi/rhi/device.h]: I removed the handlers (they were naive structs) and used std vector to manage the query pools internally in the vulkan device. Does it look better?

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

File: taichi/rhi/vulkan/vulkan_device.cpp [hunk 3]
```diff
@@ -1923,6 +1902,45 @@ Stream *VulkanDevice::get_compute_stream() {
   return iter->second.get();
 }
 
+void VulkanCommandList::begin_profiler_scope(const std::string &kernel_name) {
+  auto pool = vkapi::create_query_pool(ti_device_->vk_device());
+  vkCmdResetQueryPool(buffer_->buffer, pool->query_pool, 0, 2);
+  vkCmdWriteTimestamp(buffer_->buffer, VK_PIPELINE_STAGE_ALL_COMMANDS_BIT,
+                      pool->query_pool, 0);
+  ti_device_->profiler_add_sampler(kernel_name, pool);
```
Comment (bobcao3) [inline taichi/rhi/vulkan/vulkan_device.cpp:1910]: nit: also use this function to add debug markers for tools like renderdoc or NSight.

https://www.saschawillems.de/blog/2016/05/28/tutorial-on-using-vulkans-vk_ext_debug_marker-with-renderdoc/

**Review by bobcao3 — APPROVED** [review]
LGTM!

File: taichi/rhi/vulkan/vulkan_device.cpp (hunk 3 above)
Comment (turbo0628) [reply taichi/rhi/vulkan/vulkan_device.cpp:1910]: Thanks, we can do this in a follow-up PR

**Pull request merged**